File: Un peu d'ordre/ordre.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define MAX_VALUE 1000

typedef struct couple {
    int begin;
    int end;
} couple;

int compare(const void* a, const void* b)
{
    return (*(int*)a - *(int*)b);
}

char* order(int k, int n, int* sizes)
{
    int* sizes_sorted = malloc(n * sizeof(int));
    memcpy(sizes_sorted, sizes, n * sizeof(int));
    qsort(sizes_sorted, n, sizeof(int), compare);

    couple* mapping = malloc(MAX_VALUE * sizeof(couple));
    for (int i = 0; i < n;) {
        int end = i+1;
        while (end < n && sizes_sorted[end] == sizes_sorted[i])
            end++;
        mapping[sizes_sorted[i]] = (couple){.begin = i, .end = end};
        i = end;
    }

    for (int i = 0; i < n; i++) {
        couple wanted = mapping[sizes[i]];
        bool swap = false;
        for (int pos = wanted.begin; pos < wanted.end; pos++) {
            if ((pos - i + k) % k == 0)
                swap = true;
        }
        if (!swap) {
            free(sizes_sorted);
            free(mapping);
            return "NON";
        }
    }
    free(sizes_sorted);
    free(mapping);
    return "OUI";
}

int main(void)
{
    int k, n;
    scanf("%d %d", &k, &n);
    int* sizes = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++)
        scanf("%d", &sizes[i]);

    printf("%s\n", order(k, n, sizes));
    free(sizes);
    return 0;
}

File: Un peu d'ordre/ordre.py
from typing import List

MAX_VALUE = 1000

def counting_sort(sizes: List[int], n: int) -> List[int]:
    counting = [0] * MAX_VALUE
    for i in range(n):
        counting[sizes[i]] += 1

    sorted = []
    for i in range(MAX_VALUE):
        sorted.extend([i] * counting[i])
    return sorted

def order(k: int, n: int, sizes: List[int]) -> None:
    """
    :param k: le nombre magique
    :param n: le nombre de personnes
    :param sizes: la liste des tailles de chaque personne
    """
    mapping = sorted(sizes)
    mapping_dict = {sizes[i]: [] for i in range(n)}
    for i in range(n):
        mapping_dict[mapping[i]].append(i)

    for i in range(n):
        wanted_pos = mapping_dict[sizes[i]]
        swap = False
        for index in wanted_pos:
            if (index - i) % k == 0:
                swap = True
        if (not swap):
            return "NON"
        
    return "OUI"


if __name__ == "__main__":
    k = int(input())
    n = int(input())
    sizes = list(map(int, input().split()))
    print(order(k, n, sizes))
